File: 16-3-2024_Contiguous Array.py
from collections import defaultdict

class Solution:
    def findMaxLength(self, nums):
        n = len(nums)
        mp = defaultdict(int)
        sum = 0
        subArray_length = 0
        for i in range(n):
            sum += -1 if nums[i] == 0 else 1
            if sum == 0:
                subArray_length = max(subArray_length, i + 1)
            elif sum in mp:
                subArray_length = max(subArray_length, i - mp[sum])
            else:
                mp[sum] = i
        return subArray_length

File: 16-3-2024_Contiguous Array.java
import java.util.HashMap;

public class Solution {
    public int findMaxLength(int[] nums) {
        int n = nums.length;
        HashMap<Integer, Integer> mp = new HashMap<>();
        int sum = 0;
        int subArray_length = 0;
        for (int i = 0; i < n; i++) {
            sum += nums[i] == 0 ? -1 : 1;
            if (sum == 0) {
                subArray_length = i + 1;
            } else if (mp.containsKey(sum)) {
                subArray_length = Math.max(subArray_length, i - mp.get(sum));
            } else {
                mp.put(sum, i);
            }
        }
        return subArray_length;
    }
}


